Regenerate Resources SDKs and update Management Groups Tests

rajshah11 opened the pull request:
<!-- DO NOT DELETE THIS TEMPLATE -->

## Description
<!--
Please add an informative description that covers that changes made by the pull request.

If you are regenerating your SDK based off of a new swagger spec, please add the link to the corresponding swagger spec pull request that has been merged in the azure-rest-api-specs repository
-->
**Changes Made** 
- Regenerate SDKs in the Resources folder to match the latest specs
- Update ManagementGroups Tests 

**Management Groups Specs PRs**

[https://github.com/Azure/azure-rest-api-specs/pull/2094](https://github.com/Azure/azure-rest-api-specs/pull/2094)
[https://github.com/Azure/azure-rest-api-specs/pull/2357](https://github.com/Azure/azure-rest-api-specs/pull/2357)
[https://github.com/Azure/azure-rest-api-specs/pull/2374](https://github.com/Azure/azure-rest-api-specs/pull/2374)
[https://github.com/Azure/azure-rest-api-specs/pull/2381](https://github.com/Azure/azure-rest-api-specs/pull/2381)

---

This checklist is used to make sure that common guidelines for a pull request are followed.
- [x] Please add REST spec PR link to the SDK PR
- [x] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-net/blob/AutoRest/.github/CONTRIBUTING.md).**
- [x] **The pull request does not introduce [breaking changes](https://github.com/Azure/azure-sdk-for-net/blob/AutoRest/Documentation/breaking-changes.md).**

### [General Guidelines](https://github.com/Azure/azure-sdk-for-net/blob/AutoRest/.github/CONTRIBUTING.md#general-guidelines)
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/dev/documentation/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-net/blob/AutoRest/.github/CONTRIBUTING.md#testing-guidelines)
- [x] Pull request includes test coverage for the included changes.

### [SDK Generation Guidelines](https://github.com/Azure/azure-sdk-for-net/blob/AutoRest/.github/CONTRIBUTING.md#sdk-generation-guidelines)
- [x] If an SDK is being regenerated based on a new swagger spec, a link to the pull request containing these swagger spec changes has been included above.
- [x] The generate.cmd file for the SDK has been updated with the version of AutoRest, as well as the commitid of your swagger spec or link to the swagger spec, used to generate the code.
- [x] The `*.csproj` and `AssemblyInfo.cs` files have been updated with the new version of the SDK.

rajshah11 committed:
Regenerate SDKs

rajshah11 committed:
Regenerate SDKs to match the latest specs

rajshah11 committed:
Re-generate SDKs and update Management Group Tests

vivsriaus commented on src/SDKs/Resource/Management.ResourceManager/Generated/DeploymentsOperations.cs:
```diff
@@ -1841,7 +1841,7 @@ internal DeploymentsOperations(ResourceManagementClient client)
             if(parameters != null)
             {
-                _requestContent = SafeJsonConvert.SerializeObject(parameters, Client.SerializationSettings);
+                _requestContent = Rest.Serialization.SafeJsonConvert.SerializeObject(parameters, Client.SerializationSettings);
```
Please revert. We have our own [SerializeObject ](https://github.com/Azure/azure-sdk-for-net/blob/psSdkJson6/src/SDKs/Resource/Management.ResourceManager/Utility/SafeJsonConvertWrapper.cs) which needs to be used instead

rajshah11 replied:
Done

vivsriaus commented on src/SDKs/Resource/Management.ResourceManager/Generated/Models/Page1.cs:
```diff
@@ -0,0 +1,53 @@
+    /// </summary>
+    /// <typeparam name="T">Type of the page content items</typeparam>
+    [JsonObject]
+    public class Page1<T> : IPage<T>
```
this is not needed

vivsriaus replied:
this has still not been addressed. There shouldn't be a Page1 class

vivsriaus commented on src/SDKs/Resource/Management.ResourceManager/Generated/SdkInfo_FeatureClient.cs:
```diff
@@ -0,0 +1,18 @@
+using System.Collections.Generic;
+using System.Linq;
+
+internal static partial class SdkInfo
```
What are all these SdkInfo* classes? @shahabhijeet

shahabhijeet replied:
@vivsriaus these are generated to get more meta data about the underlying API version

vivsriaus commented on src/SDKs/Resource/Management.ResourceManager/Generated/SubscriptionsOperations.cs:
```diff
@@ -573,7 +573,7 @@ internal SubscriptionsOperations(SubscriptionClient client)
                 try
                 {
-                    _result.Body = Rest.Serialization.SafeJsonConvert.DeserializeObject<Page<Subscription>>(_responseContent, Client.DeserializationSettings);
+                    _result.Body = Rest.Serialization.SafeJsonConvert.DeserializeObject<Page1<Subscription>>(_responseContent, Client.DeserializationSettings);
```
autorest issue? Should be Page and not Page1

shahabhijeet replied:
Can you point me to the operation in the swagger for your Get?
This might not be autorest issue. This results when you slight change in your response model

vivsriaus commented on src/SDKs/Resource/Resource.Tests/InMemoryTests/DeploymentTests.InMemory.cs:
```diff
@@ -135,7 +135,7 @@ public void DeploymentTestsCreateValidateMessage()
         }
 
-        [Fact]
+        [Fact(Skip = "Failing Deployment Test")]
```
You don't have to skip this, if you revert the SerializeObject call in DeploymentOperations (see my comment there)

rajshah11 replied:
Done. Thanks

vivsriaus reviewed (changes requested):
Can you include just the management group changes and ignore changes to all other files, so you don't have to worry about skipping the tests? One more generic question: Should management groups API be a part of ARM module? Who owns the APIs?

shahabhijeet committed:
Merge branch 'psSdkJson6' into NetSdkForManagementGroups

rajshah11 commented:
@vivsriaus @shahabhijeet I think the changes suggested are all in SDKs other than ManagementGroups. I can go ahead and make the changes or just include changes related to ManagmentGroups as @vivsriaus suggested. Please let me know how should I go about it.

@vivsriaus Answering your question - We own the API but because of the nature of the service ARM suggested us to include the swagger in the same namespace and hence, the SDK too.

shahabhijeet committed:
Merge branch 'psSdkJson6' into NetSdkForManagementGroups

shahabhijeet commented:
@rajshah11 this PR cannot be merged unless @vivsriaus signs off on the concerns raised.

rajshah11 committed:
Changes to JsonConvert mentioned in the PR

rajshah11 committed:
Merge branch 'NetSdkForManagementGroups' of https://github.com/rajshah11/azure-sdk-for-net into NetSdkForManagementGroups

rajshah11 commented:
@vivsriaus @shahabhijeet I have gone ahead and made changes to JsonConvert and removed the skip flag from one of the tests as mentioned in the PR. I think you are still discussing the changes related to the "Page1" class. Can you please look into this as soon as possible? We are on a tight schedule to release,

vivsriaus commented on src/SDKs/Resource/Management.ResourceManager/Generated/Models/CreateGroupRequest.cs:
```diff
@@ -0,0 +1,64 @@
+    /// <summary>
+    /// Management group creation parameters.
+    /// </summary>
+    public partial class CreateGroupRequest
```
CreateGroup is too generic IMHO, people may confuse this to resource groups. Can you please make this CreateManagementGroupRequest instead?

rajshah11 replied:
Our swagger is already approved and the pyhton sdk is out too. Maybe we can change this in the future releases.

vivsriaus replied:
That shouldn't have been done. There are commands in CLI under az named "group", that does resource group operations. This will definitely cause confusion among our SDK users who want to create resource groups. @ravbhatnagar to comment

vivsriaus commented:
@shahabhijeet [Here](https://github.com/Azure/azure-rest-api-specs/blob/master/specification/subscription/resource-manager/Microsoft.Subscription/preview/2017-11-01-preview/subscriptionDefinitions.json#L20) is our subscriptions spec (regarding the Page->Page1 issue).

shahabhijeet commented:
@rajshah11 make sure you build at least once using this
msbuild build.proj /t:Build /p:Scope=SDKs\Resource this will update the meta data files for API version changes.

rajshah11 commented:
@shahabhijeet Thanks for the response!

1.  I did build using `msbuild build.proj /t:Build /p:Scope=SDKs\Resource` but it didn't update any files (when compared to the ones added in the previous commits). Please let me know if I am missing something.
2. [Here](https://github.com/Azure/azure-rest-api-specs/blob/master/specification/resources/resource-manager/Microsoft.Management/preview/2017-11-01-preview/management.json) is the spec for ManagementGroups. The two `x-ms-pageable` have `nextLink` and there are none without it. Do I need to change anything?

rajshah11 committed:
Props file after build

shahabhijeet reviewed (changes requested):
With all these changes you will need
1) Bump up major version (as this PR introduced braking changes in the public surface area)
2) Unless you update a version number, this will not be able publish in it's current state.

So if you are ok with breaking changes, change your
Nuget version + AssemblyFileVersion + AssemblyVersion
If not introducing breaking change, update
Nuget version + AssemblyFileVersion

vivsriaus commented on src/SDKs/Resource/Management.ResourceManager/Generated/TenantsOperations.cs:
```diff
@@ -204,7 +204,7 @@ internal TenantsOperations(SubscriptionClient client)
                 try
                 {
-                    _result.Body = Rest.Serialization.SafeJsonConvert.DeserializeObject<Page<TenantIdDescription>>(_responseContent, Client.DeserializationSettings);
+                    _result.Body = Rest.Serialization.SafeJsonConvert.DeserializeObject<Page1<TenantIdDescription>>(_responseContent, Client.DeserializationSettings);
```
Revert to Page

rajshah11 replied:
Done. Thanks!

vivsriaus reviewed (changes requested):
Page1 shouldn't be included, createGroup is ambiguos

rajshah11 committed:
Revert to Page and remove Page1

rajshah11 commented:
@vivsriaus I have reverted all Page1 references to Page and removed the Page1 class since it was identical to Page.

@vivsriaus @shahabhijeet Is it okay if I manually refactor "CreateGroupRequest" to "CreateManagementGroupRequest" without a spec update?

shahabhijeet commented:
@rajshah11 new spec and new sdk will be created. So closing this PR.
